### PR TITLE
Feature/create card view #128

### LINKED ISSUE #128: Create card/grid view for generic entities
Create a component to visualise different entities in a grid/card view. This component will be in the ```datagateway-dataview``` package.

This card view needs to be paginated. 

We will customise it for specific entities such as investigations and datasets (and there will be separate issues for those). 

This relates to https://github.com/ral-facilities/datagateway-userstories/issues/110

Acceptance criteria
- [x] A generic card view for entities supporting pagination.

### LINKED ISSUE #395: UI improvement: add pagination at the top too
**Description**:
Currently, pagination is only available at the bottom of the page. It would be helpful to add it at the top as well in the card view (below the search box). See screenshot.

Screenshot 2020-09-02 at 09 13 28

**Acceptance criteria**:
- [x] Pagination added below the search box in card view

### PR DESCRIPTION
**Pull request by GoelBiju**
## Description
Create a generic **CardView** component to show data within Cards in a Grid for varying entities (investigations/datasets).

Additionally, implements an *InvestigationCardView*/*DatasetCardView* to illustrate usage of the CardView component.

The pull request adds in the following ISIS card views:

- instruments,
- facility cycles,
- investigations,
- datasets.

The pull request adds in the following DLS (DIAMOND) card views:

- proposals,
- visits,
- datasets.

## Testing instructions

- [ ] Review code
- [ ] Check Travis build
- [ ] Review changes to test coverage

## Agile board tracking
Closes #128, closes #395

### CODE REVIEW

**Comment by louise-davies** [comment]
Comment from Tim during code review meeting: investigate whether having the header (e.g. breadcrumbs) as "sticky" and the cards scroll is independent from the page scroll is desirable

**Comment by louise-davies** [comment]
Max results randomly disappeared after clearing a filter?

**Comment by louise-davies** [comment]
Me: Certain components in cardView could be extracted out - e.g. advanced search, filter panel etc. to simplify the cardView component.

**Comment by GoelBiju** [comment]
> Comment from Tim during code review meeting: investigate whether having the header (e.g. breadcrumbs) as "sticky" and the cards scroll is independent from the page scroll is desirable

@louise-davies I am currently adding in a sticky navbar. For making the card scroll independent from the page scroll, @tswinb did you mean that the cards will be on a separate scrollable area where the filters and pagination would still be viewable without having to scroll up or down?

**Comment by GoelBiju** [comment]
> Cards per page randomly disappeared after clearing a filter?

I'll have a look into this and see if this occurs again, it as been noted [on the issue](https://github.com/ral-facilities/datagateway/issues/128#issuecomment-661770283).

**Comment by GoelBiju** [comment]
> Me: Certain components in cardView could be extracted out - e.g. advanced search, filter panel etc. to simplify the cardView component.

This has been noted on the issue and I'll look into separating parts of the CardView into separate components.

**Comment by louise-davies** [comment]
Final handover comments:
Goel: Issue to do with duplicate data due to multiple requests being sent & data being stale
Louise: This might be solved with timestamps similarly to table - we'll have to investigate

Goel: back & forward buttons work a little weirdly with filters, e.g. not reloading
Louise: I think we should wait until users complain to fix this

Goel: some tests need finishing, e2e still need doing - will make empty test cases w/ TODOs

Goel: if filter is applied beforehand, filter expanded box isn't auto expanded

Goel: upon decrease browser width, and card view going into single col view, the sort & filter box isn't full width

Goel: potentially different UI if no data available
Louise: similar to table - it just shows an empty table, no other message to user.

@GoelBiju do these issues look right?

**Comment by GoelBiju** [comment]
> Final handover comments:
> Goel: Issue to do with duplicate data due to multiple requests being sent & data being stale
> Louise: This might be solved with timestamps similarly to table - we'll have to investigate
> 
> Goel: back & forward buttons work a little weirdly with filters, e.g. not reloading
> Louise: I think we should wait until users complain to fix this
> 
> Goel: some tests need finishing, e2e still need doing - will make empty test cases w/ TODOs
> 
> Goel: if filter is applied beforehand, filter expanded box isn't auto expanded
> 
> Goel: upon decrease browser width, and card view going into single col view, the sort & filter box isn't full width
> 
> Goel: potentially different UI if no data available
> Louise: similar to table - it just shows an empty table, no other message to user.
> 
> @GoelBiju do these issues look right?

@louise-davies they all look fine, I have just discovered a small issue. If you manually specify a page in the URL query which doesn't exist for pagination it will load no data. This could be easily fixed I think, I'll have a look at it.

**Comment by GoelBiju** [comment]
> > Final handover comments:
> > Goel: Issue to do with duplicate data due to multiple requests being sent & data being stale
> > Louise: This might be solved with timestamps similarly to table - we'll have to investigate
> > Goel: back & forward buttons work a little weirdly with filters, e.g. not reloading
> > Louise: I think we should wait until users complain to fix this
> > Goel: some tests need finishing, e2e still need doing - will make empty test cases w/ TODOs
> > Goel: if filter is applied beforehand, filter expanded box isn't auto expanded
> > Goel: upon decrease browser width, and card view going into single col view, the sort & filter box isn't full width
> > Goel: potentially different UI if no data available
> > Louise: similar to table - it just shows an empty table, no other message to user.
> > @GoelBiju do these issues look right?
> 
> @louise-davies they all look fine, I have just discovered a small issue. If you manually specify a page in the URL query which doesn't exist for pagination it will load no data. This could be easily fixed I think, I'll have a look at it. I'll add it to your comment as well.

@louise-davies I have just addressed this issue although it does do a duplicate data request since the data is already loaded by the time the check for the invalid page query parameter is done.

**Comment by GoelBiju** [comment]
@louise-davies I will be fixing the currently failing tests and also adding in the test cases required for CardView.

File: packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx:29]: In several other places (e.g. `datasetDetailsPanel`) `detailsPanelResize`was made optional. It seems strange that this was altered but not made optional like elsewhere. There may be a good reason for this, but if not then maybe change this to be in line with the other detail panels?

File: packages/datagateway-download/cypress/integration/downloadCart.spec.ts
Comment (patrick-austin) [inline packages/datagateway-download/cypress/integration/downloadCart.spec.ts:140]: If these tests work, remove comment and keep change from `it.only` to `it`. Otherwise, check why we were using `only` here.

**Review by patrick-austin — CHANGES_REQUESTED** [review]
Started reviewing files, haven't got to the big changes yet.
Common and dataview unit tests passing, download failing (didn't think there were any changes to the code here?).
Linter is failing, looks like translated string (`TFunctionResult`) has the potential to be undefined, which the `title` of `BlackTooltip` doesn't accept.

File: packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx:29]: Hi @patrick-austin , the `detailsPanelResize` has not been made optional in datefile tables since there was the need to show information for each datafile in the table and a details panel would be mandatory in that case. It was previously mandatory for all tables but some details panel have been made optional so it is easier to create the card views as details panels aren't always required.

File: packages/datagateway-download/cypress/integration/downloadCart.spec.ts
Comment (GoelBiju) [reply packages/datagateway-download/cypress/integration/downloadCart.spec.ts]: @patrick-austin I think this should not have been `only`, it seems it has been there maybe from testing.

File: packages/datagateway-dataview/src/views/card/advancedFilter.component.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/card/advancedFilter.component.tsx:96]: Should use translation for these strings. Also given the removal of the search box (#418), it might make more sense to refer to it as "advanced filters" rather than "advanced search" (as there isn't a regular search)?

File: packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx
Comment (patrick-austin) [reply packages/datagateway-dataview/src/views/detailsPanels/isis/datafileDetailsPanel.component.tsx:29]: That makes sense, thanks for the explanation.

File: packages/datagateway-dataview/src/views/table/dls/dlsDatafilesTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/dls/dlsDatafilesTable.component.test.tsx:42]: Looking at the possible types for `dgcommon.data` it seems the test data has more in common with the `ICATEntity` than the `Download` entity, it just seems to be missing a `CREATE_TIME` entry. The tests are currently passing so it doesn't seem like this is problem as is, but it might be worth adding that in here (and the other tests) to surpress the warning.

File: packages/datagateway-dataview/src/views/table/dls/dlsProposalsTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/dls/dlsProposalsTable.component.test.tsx:40]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/dls/dlsVisitsTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/dls/dlsVisitsTable.component.test.tsx:43]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/isis/isisMyDataTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/isis/isisMyDataTable.component.test.tsx:44]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/isis/isisInvestigationsTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/isis/isisInvestigationsTable.component.test.tsx:44]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/isis/isisInstrumentsTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/isis/isisInstrumentsTable.component.test.tsx:42]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/isis/isisFacilityCyclesTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/isis/isisFacilityCyclesTable.component.test.tsx:40]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/isis/isisDatafilesTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/isis/isisDatafilesTable.component.test.tsx:45]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/investigationTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/investigationTable.component.test.tsx:43]: See comment on `dlsDatafilesTable.component.test.tsx`.

File: packages/datagateway-dataview/src/views/table/datafileTable.component.test.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/table/datafileTable.component.test.tsx:46]: See comment on `dlsDatafilesTable.component.test.tsx`.

**Review by patrick-austin — CHANGES_REQUESTED** [review]
Continued reviewing files, still need to look at `entityCard` and the individual card/tables for datasets, files etc.
Found some minor things I'll look into changing, as well as what to do with `pageSearch` compenent in light of #418. But as the tests are passing (albeit with low coverage), hopefully there won't be anything needing major changes .

File: packages/datagateway-dataview/src/views/card/dls/dlsDatasetsCardView.component.tsx
Comment (patrick-austin) [inline packages/datagateway-dataview/src/views/card/dls/dlsDatasetsCardView.component.tsx:1]: This, and all the ISIS specialised cardViews (`datasets`, `facilities`, `instruments`, `investigations`) have been commented out in their entiretry. It's not immediately obvious to me why, so will try commenting them back in and see if anything breaks.
If there is some more work needed here, could either do it as part of this PR or as a seperate issue. As only card view is affected for DLS/ISIS, you should still be able to access both facilities in table view and access all "generic" data in card view. But (from what I've seen on pre-prod) the page simply doesn't load the cards rather than erroring.

**Review by patrick-austin — CHANGES_REQUESTED** [review]
Finished looking over all files. A few more minor/optional changes such as removal of comments, (possibly?) redundant logic.
There's several TODOs which seem to be for non-critical changes e.g. hiding elements when they're unnecesary. Given the size of the PR it might be worth creating seperate issues for these, or documenting them on #128 (it looks like at least some of these have been recorded already) and making PRs seperately after this is merged.
A number of ISIS/DLS specific cardViews have been commented out, should try and find out why and if further work is needed decide if it should happen before or after merging this PR.
Finally there's the low test coverage (~35% for the changes). This is currently blocking merging but I imagine this could be over-ruled if we want to get this merged and write tests at a later date?
Will update this comment with as and when any decisions are made.

File: packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx
Comment (patrick-austin) [reply packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx:349]: `filterData` and `query` are only used in card view, so it make senses not to include in `clearTable`. The `query` gets preserved by the `savedView` object and `filterData` is distinct from `filters` so isn't affected by table view.
One thing I did notice is that filtering/sorting that is a shared option between card/table is preserved for each view independently. For example we can sort the table ascending, and the cards descending which will then remain in place when swapping back and forth. This seems counter intuitive to me but might be intended?

File: packages/datagateway-dataview/src/page/pageSearch.component.tsx
Comment (louise-davies) [reply packages/datagateway-dataview/src/page/pageSearch.component.tsx:22]: We don't want to remove the file, since the reason it's removed from dataview is more that we can't get this working yet, not that we'll never want the component.

File: packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx
Comment (GoelBiju) [reply packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx:327]: @patrick-austin you are right, this is quite confusing. This should have been changed earlier.

File: packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx
Comment (GoelBiju) [reply packages/datagateway-common/src/state/reducers/dgcommon.reducer.tsx:349]: @patrick-austin . at the time I think I did consider this briefly but I think the idea was whatever queries one view would have, it would belong with that view. It would make sense for the same queries to be used on either view since you are only alternating between views and not different information, but I think the issue at that time was whether the queries would work on both views. 

At the moment, like you said, the table view and card view queries are separate but as long as the table view queries do work completely alongside the card view, the views do not need to be swapped.

File: packages/datagateway-dataview/src/views/table/dls/dlsDatafilesTable.component.test.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/table/dls/dlsDatafilesTable.component.test.tsx:42]: @patrick-austin this was a warning that I had when doing tests. It didn't affect how the tests run but was more of a visual warning. It is mainly due to the fact that `ICATEntity` encompasses all the other types since this was required for how the card view would deal with different types of data.

File: packages/datagateway-dataview/src/views/card/entityCard.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/entityCard.component.tsx:211]: @patrick-austin this was just done so that we didn't have to deal with an `undefined` when it comes to rendering the title. It would just give the title in the event it is present.

File: packages/datagateway-dataview/src/views/card/cardView.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/cardView.component.tsx:468]: @patrick-austin this is one of the main items to fix with the card view. It is really an issue only because of how the different `useEffects` and updated props are interacting with each other.

File: packages/datagateway-dataview/src/views/card/cardView.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/cardView.component.tsx:296]: @patrick-austin this could be a separate issue since the idea here was to allow for the filter panel to open with the selected items already selected. At the moment due to the multiple renders going on, the item gets selected in the first render but any subsequent renders does not have the filter panel open.

File: packages/datagateway-dataview/src/views/card/datasetCardView.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/datasetCardView.component.tsx:186]: @patrick-austin this might not need to be an issue but instead, the different types of buttons can be made into separate components to be re-used.

File: packages/datagateway-dataview/src/views/card/datasetCardView.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/datasetCardView.component.tsx:128]: @patrick-austin it might be worth checking the areas in which `React.useMemo` would be appropriate over `React.useEffect`.

File: packages/datagateway-dataview/src/views/card/dls/dlsDatasetsCardView.component.tsx
Comment (GoelBiju) [reply packages/datagateway-dataview/src/views/card/dls/dlsDatasetsCardView.component.tsx:1]: @patrick-austin this was commented out since I had modified the interface and all of them needed to have been updated. They should work when uncommented.

**Comment by patrick-austin** [comment]
Have re-enabled the specialist (ISIS/DLS) views that had been commented out, and updated their state/actions in line with the generic views.

Manually testing these they seem OK, but I did notice some issues with sorting in both table and card view (#460). In part this might just be to due to missing fields on the mock data; I'm not sure if this is a known or pre-existing problem but I couldn't see an issue for it.

Everything else relating to cardView will be covered in seperate issues/PRs (e.g. #458, #459), so that this branch can be merged as it stands by @louise-davies.

And thank you @GoelBiju for giving explanations on some of my review comments, I appreciate the insight.

**Comment by GoelBiju** [comment]
> Have re-enabled the specialist (ISIS/DLS) views that had been commented out, and updated their state/actions in line with the generic views.
> 
> Manually testing these they seem OK, but I did notice some issues with sorting in both table and card view (#460). In part this might just be to due to missing fields on the mock data; I'm not sure if this is a known or pre-existing problem but I couldn't see an issue for it.
> 
> Everything else relating to cardView will be covered in seperate issues/PRs (e.g. #458, #459), so that this branch can be merged as it stands by @louise-davies.
> 
> And thank you @GoelBiju for giving explanations on some of my review comments, I appreciate the insight.

@patrick-austin it's really great that this branch is more organised now and the issues in the branch have been separated into dedicated issues in the repository! The sorting issues you mentioned in table is known by @louise-davies, but I don't think there was any formal issue created for it in the repository.

I am more than happy to help/comment when I can, please do just mention if there are any issues/queries 😃

**Review by louise-davies — APPROVED** [review]

**Pull request merged**
